tkt-80697: Add NAT support to iocage

skarekrow opened the pull request:
This supports both ipfw and pf backends, iocage will take over the firewall backend you choose, so make sure you aren't relying on it as a firewall.

Usage:
iocage set nat=1 JAIL

You can supply nat_forwards in the following configurations:

PORT <-- Will be assumed to be TCP
PROTO(PORT)
PROTO(PORT:EXTERNALPORT)
PROTO(PORTSTART-PORTEND) <-- Including pf, this will be internally replaced to the correct syntax, as otherwise it will conflict with external portmapping

In this commit:
- Common method to grab all used IP's
- Common method to generate nat IP's
- NAT support for all iocage network operations (pkglist, plugins, etc)
- Fix clean -j not correctly stopping jails
- Add new properties: nat, nat_prefix, nat_interface, nat_backend and nat_forwards
- Add NAT tests and new --nat flag for it
- Fix tests with pkg fetching and lack of allow_raw_sockets
- Use the correct exception for NotImplementedError

Ticket: #80697

Commit:
Add NAT support to iocage

This supports both ipfw and pf backends, iocage will take over the firewall backend you choose, so make sure you aren't relying on it as a firewall.

Usage:
iocage set nat=1 JAIL

You can supply nat_forwards in the following configurations:

PROTO(PORT)
PROTO(PORT:EXTERNALPORT)
PROTO(PORTSTART-PORTEND) <-- Including pf, this will be internally replaced to the correct syntax, as otherwise it will conflict with external portmapping

In this commit:
- Common method to grab all used IP's
- Common method to generate nat IP's
- NAT support for all iocage network operations (pkglist, plugins, etc)
- Fix clean -j not correctly stopping jails
- Add new properties: nat, nat_prefix, nat_interface, nat_backend and nat_forwards
- Add NAT tests and new --nat flag for it
- Fix tests with pkg fetching and lack of allow_raw_sockets
- Use the correct exception for NotImplementedError

Ticket: #80697

Commit:
Travis

sonicaj commented on iocage_lib/ioc_common.py:
```diff
@@ -957,3 +960,62 @@ def gen_unused_lo_ip():
+            )
+            pair = [_ip.exploded for _ip in network.hosts()]
+
+            if any(x in pair for x in inuse):
```
Not necessary - just a suggestion - Using set intersection ( & ) might give us more speed here

skarekrow replied:
That may be less pythonic and clear though, so I went with the clearer approach

sonicaj commented on iocage_lib/ioc_json.py:
```diff
@@ -2292,7 +2310,12 @@ def json_check_prop(self, key, value, conf):
-            'localhost_ip': ('string', )
+            'localhost_ip': ('string', ),
+            'nat': truth_variations,
+            'nat_prefix': ('string', ),
```
We should probably add validation for `nat_prefix`

sonicaj reviewed (commented):
This is awesome! Good work @skarekrow ;)
I just have a couple of minor suggestions.

william-gr commented on iocage_lib/ioc_json.py:
```diff
@@ -2458,7 +2492,7 @@ def json_check_prop(self, key, value, conf):
                 elif key == 'cpuset':
                     IOCCpuset.validate_cpuset_prop(value)
-                elif key == 'localhost_ip':
+                elif key in ('localhost_ip'):
```
This is not right, you need a comma to not make it a singleton

william-gr commented on iocage_lib/ioc_start.py:
```diff
@@ -1245,3 +1305,160 @@ def find_bridge_mtu(self, bridge):
+        if backend == 'pf':
+            self.__add_nat_pf__(nat_interface, forwards)
+            pf = su.run(
+                ['pfctl', '-f', '/tmp/iocage_nat_pf.conf'], capture_output=True
```
you should probably return `/tmp/iocage_nat_pf.conf` as a value from `__add_nat_pf__` to not repeat yourself. Same for ipfw

Commit:
Address reviews

- Fix localhost_ip conditional (result of development)
- Validate default properties being set
- Validate nat_prefix (172.16 is valid, 172.16.3 is not)
- Return the nat configuration paths for easier reuse
- Correctly capture stderr with invoke_cli
- Correctly skip empty_jail when doing restart ALL

Commit:
capture_output is 3.7 only

Commit:
These should be changed as well

sonicaj reviewed (dismissed):
Tests are green ;)

Commit:
Allow simpler syntax

If a user supplies just a port, it is assumed to be TCP

Commit:
Add debug logs for NAT

Commit:
Grab defaultrouter when one isn't supplied for VNET

Simplifies the users life with plugins.

Commit:
Ignore failure for template or empty jails in start, stop, restart tests

Commit:
Travis

Commit:
Fix tests for stop

skarekrow commented:
@sonicaj We should be all good (I am locally). Can you confirm and mark as passed?

sonicaj commented on iocage_lib/ioc_json.py:
```diff
@@ -2471,6 +2508,50 @@ def json_check_prop(self, key, value, conf):
+                            conf[key] = value
+                elif key == 'nat_prefix':
+                    if value != 'none':
+                        if not re.match(r'\d{1,3}\.\d{1,3}$', value):
```
This will allow all values like `999.999` too, i don't think that is desirable ?

Commit:
Fix non-vnet NAT plugins

Commit:
Load NAT kernel module, and add allow all rule to IPFW

Found by Waqar

Commit:
Fix order for OID since it won't exist

Also swap our way of doing accept all, this is cleaner.

Commit:
Try to unload ipfw to ensure default_to_accept

Commit:
Revert "Try to unload ipfw to ensure default_to_accept"

This reverts commit bcd3c9a850d2f2228f3530cb37196a508a75959a.

Commit:
Only load modules if they were not already loaded

A previous jail may have done this, saves us some time.

Commit:
Address review

Commit:
Drop regex, use ipaddress module instead

Suggestion by Waqar

sonicaj reviewed (approved):
Thank you ;)
